[Bugfix] handle alignment of encoder_seq_lens in mllama.py

tjohnson31415 opened the pull request:
Fix for the crash repro reported in [this comment](https://github.com/vllm-project/vllm/issues/10648#issuecomment-2722656307). The bug and fix are pretty similar to https://github.com/vllm-project/vllm/pull/12347 in that the problem arises in a batch of mixed text and image requests leading to the attn metadata having some lists with an element for each sequence and others with an element only for each sequence-with-images.

Another fix that came out of #10648

FIX #10648

<!--- pyml disable-next-line no-emphasis-as-heading -->

tjohnson31415 committed:
test: add another mllama regression case

Signed-off-by: Travis Johnson <tsjohnso@us.ibm.com>

tjohnson31415 committed:
fix mllama assertion check

Signed-off-by: Travis Johnson <tsjohnso@us.ibm.com>

tjohnson31415 committed:
little refactor and add CI test

Signed-off-by: Travis Johnson <tsjohnso@us.ibm.com>

tjohnson31415 commented on vllm/model_executor/models/mllama.py:
```diff
@@ -1250,6 +1250,31 @@ def _parse_and_validate_image_input(self, **kwargs: object):
+        # Because attn_metadata.encoder_seq_lens only counts the last
+        # group of images for each sample, which is used to cheat the
+        # block manager to allocate blocks for those images only.
+        # See input_processor_for_mllama() for more details.
```
`input_processor_for_mllama` has been removed and I don't see our use of the Transformers processor doing this same trick to only process the last image group. 

I'm working to undersatnd this better, but I think we either should re-implement the "cheat" or remove the extra complexity (i.e. this check comparing to num tiles and remove `kv_range_for_decode`).

heheda12345 replied:
I'm checking with the author of PR #11427 which removes the previous input processor. See the discussion here.
https://vllm-dev.slack.com/archives/C07QCGVDNUF/p1743001000770969

heheda12345 commented:
In general this PR looks good. Thanks for the fix. Can you sync this PR with main to get in recent bug fixes for mllama https://github.com/vllm-project/vllm/pull/15564 and https://github.com/vllm-project/vllm/pull/14883 and test it again?

tjohnson31415 committed:
fix: get test_mllama.py passing

Signed-off-by: Travis Johnson <tsjohnso@us.ibm.com>

tjohnson31415 commented:
@heheda12345 I rebased and updated the branch.
`tests/models/encoder_decoder/vision_language/test_mllama.py` tests all pass for me.

Ready for another look!

heheda12345 commented on vllm/model_executor/models/mllama.py:
```diff
@@ -1309,6 +1309,31 @@ def _parse_and_validate_image_input(self, **kwargs: object):
+        # Because attn_metadata.encoder_seq_lens only counts the last
+        # group of images for each sample, which is used to cheat the
+        # block manager to allocate blocks for those images only.
+        # See input_processor_for_mllama() for more details.
```
```suggestion
        # See MllamaMultiModalProcessor for more details.
```

heheda12345 commented on vllm/model_executor/models/mllama.py:
```diff
@@ -1309,6 +1309,31 @@ def _parse_and_validate_image_input(self, **kwargs: object):
+
+        # remove 0 encoder len entries for text-only requests for these
+        # assertions
+        attn_metadata_lens = [len for len in encoder_seq_lens if len > 0]
```
```suggestion
        attn_metadata_lens = [l for l in encoder_seq_lens if l > 0]
```
Please avoid using python built-in variable names

heheda12345 reviewed (commented):
LGTM in general. Only some nits.

tjohnson31415 committed:
commit review suggestions

Signed-off-by: Travis Johnson <tsjohnso@us.ibm.com>

tjohnson31415 commented:
@heheda12345 Thanks for the review! I pushed the suggested changes (the linter didn't like `l` as a var name so used `x` instead). PTAL

heheda12345 reviewed (approved):
Thanks for the bug fix!
